tests/loggen/file_reader.c

```diff
@@ -103,79 +104,227 @@ void close_file_reader(int nr_threads)
   g_free((gpointer)source);
 }
 
-static int
-parse_line(const char *line, char *host, char *program, char *pid, char **msg)
+static LogFormatType
+get_line_format(const char *line)
 {
-  const char *pos0;
-  const char *pos = line;
-  const char *end;
-  int space = skip_tokens;
-  int pid_len;
+  if (line == NULL)
+    return LOG_FORMAT_UNKNOWN;
+
+  char *pri_end = g_strstr_len(line, 5, ">");
+  if (!pri_end)
+    return LOG_FORMAT_UNKNOWN;
+
+  /* Guess what is the actual logformat. In BSD format
+   * priority field <> is followed by timestamp. This starts
+   * with upercase letter (Jan Feb Mar ...)
+   *
+   * In Syslog format the priority field is followed by
+   * the version number decimal.
+   */
+  if (*(pri_end + 1) >= 'A' && *(pri_end + 1) <= 'Z')
+    return LOG_FORMAT_RFC3164;
+  else if (*(pri_end + 1) >= '0' && *(pri_end + 1) <= '9')
+    return LOG_FORMAT_RFC5424;
+  else
+    return LOG_FORMAT_UNKNOWN;
+}
 
-  /* Find token */
-  while (space)
+/* copy max count or buffsize (which is less) number bytes from src to dest.
+ * put a closing '\0' after copy */
+static gsize
+safe_copy_n_string(gchar *dest, const gchar *src, int count, int buffsize, const gchar *element_name)
+{
+  if (!dest || !src)
     {
-      pos = strchr(pos, ' ');
+      ERROR("Invalid src or dest buffer (src=%p, dest=%p)\n", src, dest);
+      return 0;
+    }
 
-      if (!pos)
-        return -1;
-      pos++;
+  size_t copy_size = count+1 > buffsize ? buffsize-1 : count;
+  memcpy(dest, src, copy_size);
+  dest[copy_size] = '\0';
+  return copy_size;
+}
 
-      space --;
+static gsize
+safe_copy_string(gchar *dest, const gchar *src, int buffsize, const gchar *element_name)
+{
+  return safe_copy_n_string(dest, src, strlen(src) > buffsize ? buffsize : strlen(src), buffsize, element_name);
```

Comment:
We can delay calculating the minimum in safe_copy_n_string. That would give us a truncate message when necessary.